### DISCUSSION TITLE
dos emulator or virtual machine?

### DISCUSSION THREAD

**Discussion by stsp** [dosemu2/dosemu2#1577] [Ideas]
dosemu1 has this description:
```
DOSEMU is a user-level program which uses certain special features
of the Linux kernel and the 80386 processor to run MS-DOS/FreeDOS/
DR-DOS in what we in the biz call a `DOS box.'
```

dosemu2-pre has this:
```
DOSEMU2 is a virtual machine that allows you to run DOS software under linux.
```

But some packagers just call us
"DOS emulator version 2". :)
When dosemu2 was forked from dosemu1,
it was clearly a virtual machine to run
a bunch of real-mode OSes. So I updated
the description accordingly.
But now we have fdpp and coopthreads,
which is not a VM design at all.
VMs clearly distinguish between the
host and guest code, and in our model
such distinction vanishes.
So I wonder if we should re-brand.
What description would you prefer?
So far I can't find anything more "interesting"
than "DOS emulator version 2", but
maybe someone has a better idea?

**Comment by andrewbird** [comment]
This is a user facing message and technical correctness shouldn't really be paramount. I think it should have the word `emulator` in there somewhere. I think most people will instantly know what it's for then, whereas `virtual machine` perhaps takes more thinking about. It might also have something about using `hardware acceleration` for speed. 

Here's my stab, but sales blurbs aren't really my thing!
~~~
DOSEMU2 is a user-level emulator which can use hardware acceleration to run DOS programs
in a window. It can use either its own optimised DOS (FDPP), a proprietary MS-DOS / PC-DOS,
or FreeDOS to provide the DOS runtime environment. 
~~~

**Reply by stsp** [reply]
OK, taken as a long description vote.
Please provide a short description too.
We need a pair.

**Reply by andrewbird** [reply]
I just searched 'DOS' in Ubuntu's software manager, most others (Dosbox, emu2 etc) are going with `DOS emulator` with adjectives like 'simple' or 'open source' etc. I think we need to stress our unique advantage so maybe
`Hardware accelerated DOS emulator`

**Reply by stsp** [reply]
Arrgh...
Technically I don't like this
idea because that also underlines
our weakness, that we are not
portable. Plus we still do not do
eg hardware-accelerated rendering,
and overall we are not that
optimized for speed...
But probably you are
right that marketing is more
important.

**Reply by andrewbird** [reply]
I'm no expert though!

**Reply by stsp** [reply]
The problem is not being an
expert in marketing, but rather
in being an expert in optimizing
dosemu2 for speed. :) Which
so far only Bart ever did.
I am just worrying that such
a name will put a responsibility
on us to make it really fast.
Not sure if its a right moment
to advertise the acceleration,
but maybe we should still try. :)

**Reply by andrewbird** [reply]
I've never run dosbox, but how does it compare for speed with dosemu2?

**Reply by stsp** [reply]
That depends on many factors.
I recently went through all performance
regressions since pre8, but before
that, dosbox was faster than devel.
As of now:
- dosbox is faster than dosemu2 in debug build
(debug build is not for speed, but it is now very
slow because of sanitizers)
- dosbox is faster in mode-X games
- dosbox is probably faster in rendering because
it uses shader AFAIK, but this needs to be benched
- in most CPU-intensive games with high resolution,
dosemu2 is usually faster
- In old EGA/CGA games dosbox is sometimes
faster because of a better timer precision/granularity

Overall: you should try it yourself to compare.

**Reply by andrewbird** [reply]
Oh, I'd assumed that since we can use KVM or hardware vm86() we were faster. We probably shouldn't claim speed then.

> debug build is not for speed, but it is now very
slow because of sanitizers

Are PPA builds debug, if so should we avoid the sanitizers there somehow?

**Reply by stsp** [reply]
> Oh, I'd assumed that since we can use KVM or hardware vm86() we were faster.

Its not like we aren't.
But definitely not always.
Also we use threading but
no one have yet optimized
the locking to avoid contentions...
Speed is not something you
automatically get by the use
of some technology. Even
with KVM or threading you
need to optimize your code.
We only write the architecture
now. Our architecture supports
KVM and threading, so it can
be very fast... when its written
and optimized. :)

> Are PPA builds debug, if so should we avoid the sanitizers there somehow?

Good question.
PPA builds have debug info
but no sanitizers and optimization
enabled.
COPR builds appear to have
no sanitizers but optimization
disabled. This is an omission.

**Reply by stsp** [reply]
So how about something like this:
Short: "fast and secure DOS emulator"
Long: "dosemu2 is an emulator that allows to run DOS programs under linux.
It can also serve as a VM for running various real-mode [D]OSes."
Or not to bother about VM at all?

**Reply by andrewbird** [reply]
'allows to run' -> 'can run'.

**Reply by stsp** [reply]
"for running"?

**Comment by stsp** [comment]
So I ended up with:
`Summary: fast and secure DOS emulator`
```
%description
dosemu2 is an emulator for running DOS programs under linux.
It can also serve as a VM to boot various DOSes.
```

**Reply by andrewbird** [reply]
cool, looks good.
